File: leet_lint_code/lint_code/three_sum_closest.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 59.

#ifndef LEETLINTCODE_LINTCODE_THREESUM_CPP
#define LEETLINTCODE_LINTCODE_THREESUM_CPP

#include <vector>
#include <algorithm>
#include <math.h>

class ThreeSumCloset {
 public:
   int threeSumClosest(std::vector<int> &num, int target) {
     unsigned int zero = 0;
     int MAX = (~zero) >> 1;
     std::sort(num.begin(),num.end());
     int len = (int)num.size(), j, k;
     int closet = MAX;
     int distance = abs(MAX - target);
     for(int i = 0; i < len - 2; ++ i){
       j = i + 1, k = len - 1;
       while (j < k) {
         int sum = num[i] + num[j] + num[k];
         int tmp_distance = abs(sum - target);
         if (tmp_distance < distance) {
           distance = tmp_distance;
           closet = sum;
         }

         if(sum > target){
           while (num[k] == num[k - 1]) { -- k; }
           -- k;
         }else if(sum < target){
           while (num[j] == num[j + 1]) { ++ j;}
           ++ j;
         }else{
           return sum;
         }
       }
       while (num[i] == num[i + 1]) ++i;
     }
     return closet;
   }
};

#endif // LEETLINTCODE_LINTCODE_THREESUM_CPP

File: leet_lint_code/leet_code/java/ThreeSumClosest.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 16.
// Date: 2019.8.26

import java.util.Arrays;

public class ThreeSumClosest {

    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int i = 0;

        int dist = Integer.MAX_VALUE;
        int ans = 0;

        while (i < nums.length - 2) {
            if (i > 1 && nums[i] >= nums[i-1] && nums[i - 1] > target) {
                return ans;
            }

            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                int tmpDist = sum - target;

                if (tmpDist == 0) {
                    return target;
                } else if (tmpDist < 0){
                    while (left < nums.length - 1 && nums[left] == nums[left + 1]) ++ left;
                    ++ left;

                    tmpDist = -1 * tmpDist;
                    if (tmpDist < dist) {
                        ans = sum;
                        dist = tmpDist;
                    }

                } else {
                    while (right >= 1 && nums[right - 1] == nums[right]) -- right;
                    -- right;
                    if (tmpDist < dist) {
                        ans = sum;
                        dist = tmpDist;
                    }
                }
            }

            while (i < nums.length - 1 && nums[i + 1] == nums[i]) ++ i;
            ++ i;
        }

        return ans;
    }
}
